src/Debug.h

```diff
@@ -32,38 +32,25 @@ std::ostream &operator<<(std::ostream &stream, const Stmt &);
 struct LoweredFunc;
 std::ostream &operator<<(std::ostream &, const LoweredFunc &);
 
-/** For optional debugging during codegen, use the debug class as
+bool debug_is_active_impl(int verbosity, const char *file, const char *function, int line);
+#define debug_is_active(n) (::Halide::Internal::debug_is_active_impl((n), __FILE__, __FUNCTION__, __LINE__))
+
+/** For optional debugging during codegen, use the debug macro as
  * follows:
  *
- \code
- debug(verbosity) << "The expression is " << expr << "\n";
- \endcode
+ * \code
+ * debug(verbosity) << "The expression is " << expr << "\n";
+ * \endcode
  *
  * verbosity of 0 always prints, 1 should print after every major
  * stage, 2 should be used for more detail, and 3 should be used for
  * tracing everything that occurs. The verbosity with which to print
  * is determined by the value of the environment variable
  * HL_DEBUG_CODEGEN
  */
-
-class debug {
-    const bool logging;
-
-public:
-    debug(int verbosity)
-        : logging(verbosity <= debug_level()) {
-    }
-
-    template<typename T>
-    debug &operator<<(T &&x) {
-        if (logging) {
-            std::cerr << std::forward<T>(x);
-        }
-        return *this;
-    }
-
-    static int debug_level();
-};
+#define debug(n)                                     \
```

Comment:
Does this leak into `Halide.h`? If so we could have downstream breakage from code that relies on the `debug` symbol previously being namespaced (but now being effectively a global macro)... Ugh

Reply:
It does and that is a possibility. However, `debug` was always in `Internal`, which has very relaxed compatibility policies. AFAIK it's always been "use at your own risk".

Reply:
I think Steven was referring to the possibility that it was exposed in Halide.h and might have been inlined into the users code without explicitly calling it? That would break the ABI.

Reply:
> That would break the ABI.

We've never had any ABI guarantees between major versions, though? We don't even have API guarantees for `Halide::Internal`.

Reply:
Nah; defining global macros needs to be done with unlikely names to avoid breaking innocent code, e.g. imagine a client had code like this

```
static bool debug = false;  // set by command line flag, this does extra runtime debug checking
...
```

This code would compile just fine before, but now it would break, even though it's unrelated to the Halide API.

Thought: it would make this quite an invasive change, but how about we just rename `debug()` to `halide_debug()` everywhere?

Reply:
Why not give it the same treatment as `user_error` and the like and just `#undef` it in `Halide.h`?

Reply:
If that leaves it out of the global namespace for downstream clients, :thumbsup:

Reply:
> If that leaves it out of the global namespace for downstream clients, 👍

Done!